cobalt/devinfra/kokoro/bin/dind_build.sh

```diff
@@ -42,17 +42,28 @@ fi
 # Load common routines (we only need run_gn and ninja_build).
 . $(dirname "$0")/common.sh
 
-# Using repository root as work directory.
 WORKSPACE_COBALT="${KOKORO_ARTIFACTS_DIR}/git/src"
-cd "${WORKSPACE_COBALT}"
 
-configure_environment
-
-# Build Cobalt.
 pipeline () {
   local out_dir="${WORKSPACE_COBALT}/out/${TARGET_PLATFORM}_${CONFIG}"
-  run_gn "${out_dir}" "${TARGET_PLATFORM}" "${EXTRA_GN_ARGUMENTS:-}"
-  ninja_build "${out_dir}" "${TARGET}"
+  local src_dir="${KOKORO_ARTIFACTS_DIR}/chromium/src"
+  mkdir -p "${src_dir}"
+  cp -aT "${WORKSPACE_COBALT}" "${src_dir}"  # avoid CIPD invalid cross-device link
+
+  cd "${KOKORO_ARTIFACTS_DIR}/chromium"
+  git clone https://chromium.googlesource.com/chromium/tools/depot_tools.git tools/depot_tools
+  export PATH="${PATH}:${KOKORO_ARTIFACTS_DIR}/chromium/tools/depot_tools"
+  gclient config --name=src rpc://lbshell-internal/cobalt_src
+  if [[ "${TARGET_PLATFORM}" =~ "android" ]]; then
+    echo "target_os=['android']" >> .gclient
+  fi
+
+  cd "${src_dir}"
+  git config --global --add safe.directory "${src_dir}"
+  gclient sync -v --shallow --no-history -r "${KOKORO_GIT_COMMIT_src}"
+  cobalt/build/gn.py -p "${TARGET_PLATFORM}" -C "${CONFIG}"
+  ninja -C "out/${TARGET_PLATFORM}_${CONFIG}" ${TARGET}  # TARGET expands to multiple args
+  cp -a out "${WORKSPACE_COBALT}"
```

Comment:
I'm not too happy with the amount of code included here in the main pipeline function. Could you this up a into steps, similar to the old code?

Reply:
I suspect abstracting this would be worse in a number of ways.

There are about 10 operations here, all semantically related,
that need to be done almost exactly in the order they appear.
Extracting functions would tend to obscure dependencies,
mislead unfamiliar readers, and introduce [sequential coupling](https://fermibubble.github.io/anti-patterns/sequential-coupling).

While developing this code I did consider how to organize it
and this seems good to me, but specific suggestions are welcome.

Reply:
I think my complaint was based on the fact that this code was previously organized as similarly to the GitHub build pipeline as was possible in bash whereas now looks more like a single step (or possibly two) expanded in the main "pipeline" code. I also don't fully agree with the argument about sequential coupling here. We're not designing an API to be used throughout a code base but rather splitting a build up into a series of steps with each step consisting of a series of commands.

I can identify ~4 distinct build steps here. Roughly: directory setup, gclient setup and sync, gn generation, and ninja build with some extra sprinkles of logic interspersed. If you add some space between these and add comments before each I think that's enough.

Reply:
done against my preferences 😉